File: medium/2095_delete_the_middle_node_of_a_linked_list/solution.py
from typing import Optional

from utils.python.nodes import ListNode


class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        slow = head
        fast = head

        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next
            if fast:
                fast = fast.next

        if prev:
            prev.next = slow.next
            del slow
        else:
            head = None

        return head

File: medium/2095_delete_the_middle_node_of_a_linked_list/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  utils::ListNode *deleteMiddle(utils::ListNode *head) {
    if (!head || !head->next)
      return nullptr;

    utils::ListNode *slow = head;
    utils::ListNode *fast = head->next->next;

    while (fast && fast->next) {
      slow = slow->next;
      fast = fast->next->next;
    }

    slow->next = slow->next->next;

    return head;
  }
};
